hpms/src/main/java/gov/cms/ab2d/hpms/service/AttestationUpdaterService.java

```diff
@@ -0,0 +1,151 @@
+                .filter(hpmsInfo -> !existingMap.containsKey(hpmsInfo.getContractId()))
+                .collect(Collectors.toList());
+        List<Contract> contractAttestList = addNewContracts(newContracts);
+
+        Map<String, HPMSOrganizationInfo> refreshed = buildRefreshedMap(orgInfo);
+        existingMap.forEach((contractId, contract) ->
+                considerContract(contractAttestList, contract, refreshed.get(contractId)));
+
+        batchAttestations(contractAttestList.stream().map(Contract::getContractNumber).collect(Collectors.toList()));
+    }
+
+    // Limit the size of the request to BATCH_SIZE, avoiding URLs that are too long and keeping the burden down
+    // on the invoked service.
+    private void batchAttestations(List<String> contractAttestList) {
+        final int size = contractAttestList.size();
+        int startIdx = 0;
+        for (; startIdx < size - BATCH_SIZE; startIdx += BATCH_SIZE) {
+            List<String> currentChunk = contractAttestList.subList(startIdx, startIdx + BATCH_SIZE);
+            processAttestations(currentChunk);
+        }
+
+        // process the remainder (if any) - i.e. smaller than a batch
+        if (size % BATCH_SIZE != 0) {
+            List<String> currentChunk = contractAttestList.subList(startIdx, size);
+            processAttestations(currentChunk);
+        }
+    }
+
+    private void processAttestations(List<String> currentChunk) {
+        String contractIdStr = currentChunk.stream().collect(Collectors.joining("\",\"", "[\"", "\"]"));
+        hpmsFetcher.retrieveAttestationInfo(this::processContracts, contractIdStr);
+    }
+
+    private void processContracts(HPMSAttestationsHolder contractHolder) {
+        Map<String, Contract> existingMap = buildExistingContractMap();
+        contractHolder.getContracts()
+                .forEach(attest -> updateContractIfChanged(attest, existingMap.get(attest.getContractId())));
+    }
+
+    private void updateContractIfChanged(HPMSAttestation attest, Contract contract) {
+        if (contract.updateAttestation(attest.isAttested(), attest.getAttestationDate(), FORMATTER)) {
+            contractRepository.save(contract);
+        }
+    }
+
+    List<Contract> addNewContracts(List<HPMSOrganizationInfo> newContracts) {
+        if (newContracts.isEmpty()) {
+            return new ArrayList<>();
+        }
+        return newContracts.stream().map(this::sponserAdd).collect(Collectors.toList());
+    }
+
+    private Contract sponserAdd(HPMSOrganizationInfo hpmsInfo) {
+        Sponsor sponsor = new Sponsor();
+        sponsor.setHpmsId(hpmsInfo.getParentOrgId());
+        sponsor.setOrgName(hpmsInfo.getParentOrgName());
+        Sponsor savedSponser = sponsorRepository.save(sponsor);
+
+        Contract retContract = new Contract();
+        retContract.setContractName(hpmsInfo.getContractName());
+        retContract.setContractNumber(hpmsInfo.getContractId());
+        retContract.setSponsor(savedSponser);
```

Comment:
Based on what Kevin was saying I think this would be associating the contract with the parent org, contracts should be associated with the organization that they are apart of.

Reply:
You may have to search and see if the contract exists, and then get the sponsor based off of that

Reply:
Ok it sounds like the orgMarketingName is what may match our org name for the organization. You may want to confirm by comparing some prod data to what you'd get back from their API.

Reply:
What if the Contract does not exist?

Reply:
Then we'd want to create a new sponsor.

Reply:
I key off of Contract, not Sponsor initially: contractRepo.findAll() => existingMap.

Where things are a bit rough, is when we detect a new contract.